### ISSUE TITLE
DROPMALFORMED vs PERMISSIVE

### ISSUE THREAD

**Issue by SrikanthTati** [databricks/spark-csv#255]
I have many CSV files coming from different upstream sources. Most of the files stick to a schema.
Ex,
```
val schema = new StructType(
      Array(
        StructField("Name",StringType,true),
        StructField("Place",StringType,true),
        StructField("Age",IntegerType,true)
      )
)
```
Sample data
```
#Name,Place,Age
Bob,NY,40
Joe,SF,30
Tom,LA,25
```

When a schema change happens(new column added), some upstream sources use new schema sooner than others. So we have csv files with old and new schema. Spark workflow doesn't change as `PERMISSIVE` mode will just omit new column(until we change code to process new column).
But we had to use `DROPMALFORMED` mode as some(very few) rows were malformed. This will cause rows written using new schema to be dropped as well. Is there a way to handle this.

```
#Name,Place,Age
Bob,NY,40  --> Row in old schema
Joe,SF,30,extra  --> Row in new schema with extra column
Tom,,LA,25  --> Malformed row
```

`PERMISSIVE` doc says "tries to parse all lines: nulls are inserted for missing tokens and extra tokens are ignored.". 

https://github.com/databricks/spark-csv/blob/master/src/main/scala/com/databricks/spark/csv/CsvRelation.scala#L124

Why not handle exception too by inserting nulls similar to ArrayIndexOutOfBound?
Any suggestions on how to handle this?

**Comment by SrikanthTati** [comment]
Any thoughts on this? If this needs a code change, I can create a PR.

**Comment by falaki** [comment]
Yes, this requires changes to the code. We basically need a new parsing mode that accepts rows with missing fields.

**Comment by SrikanthTati** [comment]
Shouldn't this be part of what `permissive` mode does?

**Comment by falaki** [comment]
PERMISSIVE does that, but you need to make sure the data source parses the data with new schema. That can be done by specifying schema manually and using PERMISSIVE mode. But in that case it will also accept those few malformed rows.

**Comment by SrikanthTati** [comment]
I'm already passing in the schema with the default PERMISSIVE mode.

     sqlContext.read.format("com.databricks.spark.csv")
     .options(Map("path" -> (folder + "/*.lzo"), "delimiter" -> "\t")
     .schema(RawSchema.getSchema).load()

With this, the first row in my example is read correctly. Second row is also read by omitting the "extra" column. It's the third(malformed) row that is raising an exception because of the extra (,). PERMISSIVE doesn't handle this. Hence I was thinking this has to be fixed instead of adding a new mode.
